library/src/commonMain/kotlin/com/lagradost/cloudstream3/extractors/DoodExtractor.kt

```diff
@@ -58,31 +60,62 @@ class DoodYtExtractor : DoodLaExtractor() {
     override var mainUrl = "https://dood.yt"
 }
 
+class Ds2play : DoodLaExtractor() {
+    override var mainUrl = "https://ds2play.com"
+}
+
+class Ds2video : DoodLaExtractor() {
+    override var mainUrl = "https://ds2video.com"
+}
+
 open class DoodLaExtractor : ExtractorApi() {
     override var name = "DoodStream"
     override var mainUrl = "https://dood.la"
     override val requiresReferer = false
-
-    override fun getExtractorUrl(id: String): String {
-        return "$mainUrl/d/$id"
-    }
-
-    override suspend fun getUrl(url: String, referer: String?): List<ExtractorLink>? {
-        val newUrl= url.replace(mainUrl, "https://d0000d.com")
-        val response0 = app.get(newUrl).text // html of DoodStream page to look for /pass_md5/...
-        val md5 ="https://d0000d.com"+(Regex("/pass_md5/[^']*").find(response0)?.value ?: return null)  // get https://dood.ws/pass_md5/...
-        val trueUrl = app.get(md5, referer = newUrl).text + "zUEJeL3mUN?token=" + md5.substringAfterLast("/")   //direct link to extract  (zUEJeL3mUN is random)
-        val quality = Regex("\\d{3,4}p").find(response0.substringAfter("<title>").substringBefore("</title>"))?.groupValues?.get(0)
-        return listOf(
+	
+    private val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+
+    override suspend fun getUrl(
+        url: String,
+        referer: String?,
+        subtitleCallback: (SubtitleFile) -> Unit,
+        callback: (ExtractorLink) -> Unit
+    ) {
+		val req = app.get(url)
+        val host = getBaseUrl(req.url)
+        val response0 = req.text
+	val md5 = host + (Regex("/pass_md5/[^']*").find(response0)?.value ?: return)
+        val trueUrl = app.get(md5, referer = req.url).text + createHashTable() + "?token=" + md5.substringAfterLast("/")
+		
+	val quality = Regex("\\d{3,4}p")
+            .find(response0.substringAfter("<title>").substringBefore("</title>"))
+            ?.groupValues
+            ?.getOrNull(0)
+		
+	callback.invoke(
             ExtractorLink(
                 this.name,
                 this.name,
                 trueUrl,
-                mainUrl,
+                "$mainUrl/",
                 getQualityFromName(quality),
-                false
+                INFER_TYPE,
             )
-        ) // links are valid in 8h
+        )
 
     }
-}
\ No newline at end of file
+	
+private fun createHashTable(): String {
+        return buildString {
+            repeat(10) {
+                append(alphabet[Random.nextInt(alphabet.length)])
```

Comment:
alphabet.random() is nice to have here :+1:

Reply:
done